samcli/commands/local/cli_common/invoke_context.py

```diff
@@ -44,7 +44,8 @@ def __init__(self,
                  docker_network=None,
                  log_file=None,
                  skip_pull_image=None,
-                 aws_profile=None):
+                 aws_profile=None,
+                 delve_path=None):
```

Comment:
I read your comment about `delve_path` in the Issue. Yes, we should generalize this and not pass around one parameter specific to Golang everywhere.

Also, I think this is a good time to create a new data object say `DebugingContext` that encapsulates all properties necessary for debugging - debug port, debug args and delve path. This will allow us to first simplify the call signatures, second abstract away how customers pass debug information (CLI args, config file, sane defaults), third auto-load debug context in future from other config files like vscode's launch configuration if it has all info you need. 

On that note, I am curious if you can reuse `debug_args` property for `delve_path`? 

What do you think about these two?

Reply:
I'm leery of reusing `debug_args` for the `delve_path` because I can see end-users wanting to pass in options via that and there's no reason the container _can't_ support it. I think a `DebugContext` object makes a lot of sense though. I think the `delve_path` can be generalized to `debugger_path` and packed into the new debug context. That'll enable us to re-use the argument for bringing in `netcore` debuggers.

Reply:
Random thought but something we should be considering. I do prefer the general path but I think we need to incorporate some specific language runtime into the `debugger_path` option.

One the of the benefits of serverless, is that each Lambda can be its own runtime. While this is the invoke_context, it is also used in the [start-api command](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/commands/local/start_api/cli.py#L42). For a specific `invoke`, I think a generic `debugger_path` makes perfect sense but not sure about in the context of `start-api`. By giving a generic path, I feel that we restrict an API GW + Lambda to be one language when it doesn't need to (or should) be. If the `debug_path` was a mapping of runtime to a path, this can be easily supported.

Thoughts?

Reply:
I think that makes sense, yeah. To expand on it, what if the `DebugContext` was runtime-specific and the CLI allowed for multiple debug flags? Envisioning something like the following as an invocation:

`sam local start-api -d runtime=go1.x debug_port=5986 debugger_path=/path/to/debugger -d runtime=nodejs8.10 debug_port=5986 debug_args='--foo'...`

This would probably end up being a breaking change, but I'm not sure how big of a deal it would be; probably not a lot of people using automation around the `-d` option.

Reply:
@jfuss / @sanathkr - I'm going to plead ignorance on python here, but after exploring the Click docs for a while, I can't seem to find a way to have it deserialize named arguments to an option like I did in my invocation example. The best thing I can find would be to load the debug context object from a file. Y'all think that would work? So it'd look something like this...

`sam local start-api -d debug_contexts.json...`

debug_contexts.json
```
{
  "runtime": "go1.x",
  "port": 5986,
  "debugger_path": "path/to/dlv",
  "debug_args": ""
},
{
  "runtime": "nodejs8.10",
  "port": 5986,
  "debug_args": "--foo --bar=baz"
}
```